README.md
# Attributed

[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![CocoaPods Compatible](https://img.shields.io/cocoapods/v/DAHAttributed.svg)](https://cocoapods.org/pods/DAHAttributed)
[![GitHub license](https://img.shields.io/badge/license-MIT-lightgrey.svg)](./LICENSE.txt)
[![Build Status](https://travis-ci.org/davidahouse/Attributed.svg)](https://travis-ci.org/davidahouse/Attributed)

Creating NSAttributedString objects can be cumbersome to do in code. This library adds a helper object that can apply attributes to strings easily.

## Installation

### [Carthage]

[Carthage]: https://github.com/Carthage/Carthage

Add the following to your Cartfile:

```ruby
github "davidahouse/Attributed" "master"
```

Then run `carthage update`.

Follow the current instructions in [Carthage's README][carthage-installation]
for up to date installation instructions.

[carthage-installation]: https://github.com/Carthage/Carthage/blob/master/README.md

### [CocoaPods]

[CocoaPods]: http://cocoapods.org

Add the following to your [Podfile](http://guides.cocoapods.org/using/the-podfile.html):

```ruby
pod 'DAHAttributed'
```

You will also need to make sure you're opting into using frameworks:

```ruby
use_frameworks!
```

Then run `pod install`.

## Overview

Creating a NSAttributedString with different formats applied to different sections of the string involves
a lot of code. If you have some basic formatting needs, and some strings, this library allows for creating
an NSAttributedString quickly.

The helper class is called `Attributed` contains a set of formatting attributes, then gives a simple set
of methods for applying those attributes to strings or NSAttributedStrings. It also provides a way for
strings and NSAttributedStrings to be added together.

### Creating an `Attributed` object

- No formatting at all

```swift
let noFormatting = Attributed()
```

- Apply a color

```swift
let redText = Attributed(color: NSColor.redColor())
let blueText = Attributed(color: NSColor.blueColor())
```

- Apply a font

```swift
let boldText = Attributed(font: NSFont.boldSystemFontOfSize(24))
```

- Multiple values

```swift
let detailsText = Attributed(attributes: [NSForegroundColorAttributeName: NSColor.greenColor(), NSFontAttributeName: NSFont.systemFontOfSize(18)])
```

### Using an `Attributed` object

- Apply the formatting to a String

```swift
let stringWithAttributes = redText.toString { "Hi mom!" }
```

- Apply the formatting to a NSAttributedString

```swift
let attributedStringWithAttributes = redText.toString { NSAttributedString(string: "The quick brown fox") }
```

- Combining Strings

```swift
let multipleStrings = boldText.combine(strings: "Hi mom!", "The quick brown fox")
```

- Nesting

```swift
let nestedStrings = redText.toString { boldText.toString { "Header: " } + "here are the details" }
```

- More nesting

```swift
let anotherNestedString = redText.toString { boldText.toString { "Header: " } + detailsText.toString { "here are the details" } + blueText.toString{ "@" } }
```

- Strings and attributed strings can be added together

```swift
let addedString = stringWithAttributes + " How are you?"
```


CHANGELOG.md
# Attributed Changelog

## Version 1.1.0

- Changed the `string` method names to `toString`
- Added a `toString` extension on NSColor, UIColor, NSFont and UIFont

## Version 1.0.1

- Added a Changelog
- Added inline documentation to Attributed.swift

## Version 1.0.0

- Initial release


Sources/Attributed.swift
//
//  Attributed.swift
//  Attributed
//
//  Created by David House on 1/10/16.
//  Copyright © 2016 David House. All rights reserved.
//

import Foundation

/// A helper class for creating NSAttributedString objects
public class Attributed {
    
    let attributes: [String: AnyObject]
    
    // MARK: Initializers
    
    /// Initialize with an empty set of attributes
    public init() {
        attributes = [String: AnyObject]()
    }
    
    /// Initialize with only a foreground color attribute
    ///
    /// - parameter color: The foreground color attribute
    public init(color: AttributedColor) {
        attributes = [AttributedColorAttributeName: color]
    }
    
    /// Initialize with only a font attribute
    /// 
    /// - parameter font: The font attribute
    public init(font: AttributedFont) {
        attributes = [AttributedFontAttributeName: font]
    }
    
    /// Initialize with a dictionary of attributes
    ///
    /// - parameter attributes: A dictionary of attributed string attributes
    public init(attributes: [String: AnyObject]) {
        self.attributes = attributes
    }
    
    // MARK: String methods

    /// Create an NSAttributedString from a String,
    /// while applying attributes.
    ///
    /// - parameter inner: A String
    /// - returns: An NSAttributedString that has the attributes applied to it
    public func toString(inner: String) -> NSAttributedString {
        return NSAttributedString(string: inner, attributes: attributes)
    }
    
    /// Create an NSAttributedString from a closure that returns a String,
    /// while applying attributes.
    ///
    /// - parameter inner: A closure that returns a String
    /// - returns: An NSAttributedString that has the attributes applied to it
    public func toString(inner:() -> String) -> NSAttributedString {
        return NSAttributedString(string: inner(), attributes: attributes)
    }
    
    /// Create an NSAttributedString from an NSAttributedString, while also applying our attributes
    ///
    /// This method will not overwrite any attribute that has already been
    /// applied to the string.
    ///
    /// - parameter inner: An NSAttributedString
    /// - returns: An NSAttributedString that has the attributes applied to it
    public func toString(inner: NSAttributedString) -> NSAttributedString {
        let output = NSMutableAttributedString(attributedString: inner)
        guard output.length > 0 else {
            return output
        }
        
        var startIndex = 0
        while ( startIndex < output.length ) {
            var range = NSRange()
            let attributesAt = output.attributesAtIndex(startIndex, effectiveRange: &range)
            for attribute in attributes {
                if !attributesAt.keys.contains(attribute.0) {
                    output.addAttribute(attribute.0, value: attribute.1, range: range)
                }
            }
            startIndex = startIndex + range.length
        }
        return output
    }
    
    /// Create an NSAttributedString from a closure that returns an NSAttributedString, while also applying our attributes
    ///
    /// This method will not overwrite any attribute that has already been
    /// applied to the string.
    ///
    /// - parameter inner: A closure that returns an NSAttributedString
    /// - returns: An NSAttributedString that has the attributes applied to it
    public func toString(inner:() -> NSAttributedString) -> NSAttributedString {
        
        return toString(inner())
    }

    /// Create an NSAttributedString by combining multiple Strings using 
    /// a separator, while also applying our attributes.
    ///
    /// - parameter separator: The separator to use between each String, with space being the default
    /// - parameter strings: A variable number of String objects that will be combined
    /// - returns: An NSAttributedString with all the Strings combined and attributes applied
    public func combine(separator: String = " ", strings: String...) -> NSAttributedString {
        let output = NSMutableAttributedString()
        for inner in strings {
            if output.length > 0 {
                output.appendAttributedString(NSAttributedString(string: separator))
            }
            output.appendAttributedString(toString{ inner })
        }
        return output
    }

    /// Create an NSAttributedString by combining multiple NSAttributedStrings using
    /// a separator, while also applying our attributes.
    ///
    /// - parameter separator: The separator to use between each String, with space being the default
    /// - parameter strings: A variable number of NSAttributedString objects that will be combined
    /// - returns: An NSAttributedString with all the NSAttributedStrings combined and attributes applied
    public func combine(separator: String = " ", strings: NSAttributedString...) -> NSAttributedString {
        let output = NSMutableAttributedString()
        for inner in strings {
            if output.length > 0 {
                output.appendAttributedString(NSAttributedString(string: separator))
            }
            output.appendAttributedString(toString{ inner })
        }
        return output
    }
}

/// Add two NSAttributedStrings together
///
/// - parameter left: An NSAttributedString
/// - parameter right: Another NSAttributedString
/// - returns: An NSAttributedString representing left + right
public func + (left:NSAttributedString, right:NSAttributedString) -> NSAttributedString {
    let output = NSMutableAttributedString(attributedString: left)
    output.appendAttributedString(right)
    return output
}

/// Add an NSAttributedString and String together
///
/// - parameter left: An NSAttributedString
/// - parameter right: A String
/// - returns: An NSAttributedString representing left + right
public func + (left:NSAttributedString, right:String) -> NSAttributedString {
    let output = NSMutableAttributedString(attributedString: left)
    output.appendAttributedString(NSAttributedString(string: right))
    return output
}

/// Add a String and NSAttributedString together
///
/// - parameter left: A String
/// - parameter right: An NSAttributedString
/// - returns: An NSAttributedString representing left + right
public func + (left:String, right:NSAttributedString) -> NSAttributedString {
    let output = NSMutableAttributedString(attributedString: NSAttributedString(string: left))
    output.appendAttributedString(right)
    return output
}


Sources/Attributed+UIKit.swift
//
//  Attributed+UIKit.swift
//  Attributed
//
//  Created by David House on 4/12/16.
//  Copyright © 2016 David House. All rights reserved.
//

import UIKit

// MARK: typealias for UIKit values

public typealias AttributedColor = UIColor
public typealias AttributedFont = UIFont
public let AttributedColorAttributeName: String = NSForegroundColorAttributeName
public let AttributedFontAttributeName: String = NSFontAttributeName

// MARK: UIColor extension

extension UIColor {
    
    /// Create an NSAttributedString from a string while applying a
    /// UIColor attribute.
    ///
    /// - parameter inner: A String
    /// - returns: An NSAttributedString that has the attributes applied to it
    public func toString(innerString: String) -> NSAttributedString {
        let attributes = [NSForegroundColorAttributeName: self]
        return NSAttributedString(string: innerString, attributes: attributes)
    }
}

// MARK: NSFont extension

extension UIFont {
    
    /// Create an NSAttributedString from a string while applying a
    /// UIFont attribute.
    ///
    /// - parameter inner: A String
    /// - returns: An NSAttributedString that has the attributes applied to it
    public func toString(inner: String) -> NSAttributedString {
        let attributes = [NSFontAttributeName: self]
        return NSAttributedString(string: inner, attributes: attributes)
    }
}


Sources/Attributed+Foundation.swift
//
//  Attributed+Foundation.swift
//  Attributed
//
//  Created by David House on 4/12/16.
//  Copyright © 2016 David House. All rights reserved.
//

import AppKit
import Foundation

// MARK: typealias for AppKit values

public typealias AttributedColor = NSColor
public typealias AttributedFont = NSFont
public let AttributedColorAttributeName: String = NSForegroundColorAttributeName
public let AttributedFontAttributeName: String = NSFontAttributeName

// MARK: NSColor extension

extension NSColor {
    
    /// Create an NSAttributedString from a string while applying a 
    /// NSColor attribute.
    ///
    /// - parameter inner: A String
    /// - returns: An NSAttributedString that has the attributes applied to it
    public func toString(inner: String) -> NSAttributedString {
        let attributes = [NSForegroundColorAttributeName: self]
        return NSAttributedString(string: inner, attributes: attributes)
    }    
}

// MARK: NSFont extension

extension NSFont {
    
    /// Create an NSAttributedString from a string while applying a
    /// NSFont attribute.
    ///
    /// - parameter inner: A String
    /// - returns: An NSAttributedString that has the attributes applied to it
    public func toString(inner: String) -> NSAttributedString {
        let attributes = [NSFontAttributeName: self]
        return NSAttributedString(string: inner, attributes: attributes)
    }
}


AttributedTests/AttributedTests.swift
//
//  AttributedTests.swift
//  AttributedTests
//
//  Created by David House on 1/10/16.
//  Copyright © 2016 David House. All rights reserved.
//

import XCTest
@testable import Attributed

let aString: String = "Here is a string to use in tests"

class AttributedTests: XCTestCase {

    func testCanCreateAStringWithNoFormatting() {

        let attributed = Attributed()
        let result = attributed.toString { aString }
        XCTAssertEqual(result, NSAttributedString(string: aString))
    }

    func testCanCreateAnEmptyString() {

        let attributed = Attributed()
        let result = attributed.toString { "" }
        XCTAssertEqual(result.length, 0)

        let attributedResult = attributed.toString { NSAttributedString() }
        XCTAssertEqual(attributedResult.length, 0)
    }

    func testCanCreateAStringWithOnlyColorApplied() {

        let attributed = Attributed(color: NSColor.redColor())
        let result = attributed.toString(aString)
        var range = NSRange()
        let attributes = result.attributesAtIndex(0, effectiveRange: &range)
        XCTAssertEqual(attributes.count, 1)
        XCTAssertTrue(attributes.keys.contains(NSForegroundColorAttributeName))
        XCTAssertEqual(range.location, 0)
        XCTAssertEqual(range.length, result.length)
    }

    func testCanCreateAStringWithOnlyColorAppliedUsingClosure() {

        let attributed = Attributed(color: NSColor.redColor())
        let result = attributed.toString { aString }
        var range = NSRange()
        let attributes = result.attributesAtIndex(0, effectiveRange: &range)
        XCTAssertEqual(attributes.count, 1)
        XCTAssertTrue(attributes.keys.contains(NSForegroundColorAttributeName))
        XCTAssertEqual(range.location, 0)
        XCTAssertEqual(range.length, result.length)
    }

    func testCanCreateStringWithOnlyFontApplied() {

        let attributed = Attributed(font: NSFont.boldSystemFontOfSize(24))
        let result = attributed.toString { aString }
        var range = NSRange()
        let attributes = result.attributesAtIndex(0, effectiveRange: &range)
        XCTAssertEqual(attributes.count, 1)
        XCTAssertTrue(attributes.keys.contains(NSFontAttributeName))
        let font = attributes[NSFontAttributeName] as? NSFont
        XCTAssertEqual(font, NSFont.boldSystemFontOfSize(24))
        XCTAssertEqual(range.location, 0)
        XCTAssertEqual(range.length, result.length)
    }

    func testCanCreateStringWithMultipleAttributesAtOnce() {

        let attributes = [NSForegroundColorAttributeName: NSColor.redColor(),
                          NSFontAttributeName: NSFont.boldSystemFontOfSize(24)]
        let attributed = Attributed(attributes: attributes)
        let result = attributed.toString { aString }
        var range = NSRange()
        let foundAttributes = result.attributesAtIndex(0, effectiveRange: &range)
        XCTAssertEqual(foundAttributes.count, attributes.count)
        XCTAssertEqual(foundAttributes[NSForegroundColorAttributeName] as? NSColor,
                       attributes[NSForegroundColorAttributeName] as? NSColor)
        XCTAssertEqual(foundAttributes[NSFontAttributeName] as? NSFont,
                       attributes[NSFontAttributeName] as? NSFont)
        XCTAssertEqual(range.location, 0)
        XCTAssertEqual(range.length, result.length)
    }

    func testThatAttributedStringCanFormatted() {

        let plainAttributedString = NSAttributedString(string: aString)
        let attributed = Attributed(color: NSColor.redColor())
        let result = attributed.toString { plainAttributedString }
        var range = NSRange()
        let attributes = result.attributesAtIndex(0, effectiveRange: &range)
        XCTAssertEqual(attributes.count, 1)
        XCTAssertTrue(attributes.keys.contains(NSForegroundColorAttributeName))
        XCTAssertEqual(range.location, 0)
        XCTAssertEqual(range.length, result.length)
    }

    func testThatApplyingAttributesDoesntOverrideExistingAttributes() {

        let plainAttributedString = NSAttributedString(string: aString)
        let attributed = Attributed(color: NSColor.redColor())
        let result = attributed.toString { plainAttributedString }
        var range = NSRange()
        let attributes = result.attributesAtIndex(0, effectiveRange: &range)
        XCTAssertEqual(attributes.count, 1)
        XCTAssertTrue(attributes.keys.contains(NSForegroundColorAttributeName))
        XCTAssertEqual(range.location, 0)
        XCTAssertEqual(range.length, result.length)

        let greenAttributed = Attributed(color: NSColor.greenColor())
        let finalResult = greenAttributed.toString { result }
        let finalAttributes = finalResult.attributesAtIndex(0, effectiveRange: &range)
        XCTAssertEqual(finalAttributes.count, 1)
        XCTAssertTrue(finalAttributes.keys.contains(NSForegroundColorAttributeName))
        XCTAssertEqual(finalAttributes[NSForegroundColorAttributeName] as? NSColor,
                       NSColor.redColor())
        XCTAssertEqual(range.location, 0)
        XCTAssertEqual(range.length, result.length)
    }

    func testAttributedStringsCanBeAddedTogether() {

        let firstString = NSAttributedString(string: aString)
        let secondString = NSAttributedString(string: aString)
        let result = firstString + secondString
        XCTAssertEqual(result.length, firstString.length + secondString.length)
    }

    func testAttributedAndPlainStringsCanBeAddedTogether() {

        let firstString = NSAttributedString(string: aString)
        let result = firstString + aString
        XCTAssertEqual(result.length, firstString.length + aString.characters.count)
        let resultOtherWay = aString + firstString
        XCTAssertEqual(resultOtherWay.length, firstString.length + aString.characters.count)
    }

    func testCanCombineStrings() {

        let attributed = Attributed()
        let result = attributed.combine(strings: aString, aString, aString)
        XCTAssertEqual(result.length, (aString.characters.count * 3) + 2)
    }

    func testCanCominbeAttribtedStrings() {

        let firstString = NSAttributedString(string: aString)
        let attributed = Attributed()
        let result = attributed.combine(strings: firstString, firstString, firstString)
        XCTAssertEqual(result.length, (firstString.length * 3) + 2)
    }

    func testCanCombineStringsWhileSpecifyingASeparator() {

        let attributed = Attributed()
        let result = attributed.combine("", strings: aString, aString, aString)
        XCTAssertEqual(result.length, (aString.characters.count * 3))
    }

    func testCanCreateAnAttributedStringFromANSColor() {

        let result = NSColor.redColor().toString(aString)
        var range = NSRange()
        let attributes = result.attributesAtIndex(0, effectiveRange: &range)
        XCTAssertEqual(attributes.count, 1)
        XCTAssertTrue(attributes.keys.contains(NSForegroundColorAttributeName))
        XCTAssertEqual(range.location, 0)
        XCTAssertEqual(range.length, result.length)
    }

    func testCanCreateAnAttributedStringFromANSFont() {

        let result = NSFont.boldSystemFontOfSize(24).toString(aString)
        var range = NSRange()
        let attributes = result.attributesAtIndex(0, effectiveRange: &range)
        XCTAssertEqual(attributes.count, 1)
        XCTAssertTrue(attributes.keys.contains(NSFontAttributeName))
        let font = attributes[NSFontAttributeName] as? NSFont
        XCTAssertEqual(font, NSFont.boldSystemFontOfSize(24))
        XCTAssertEqual(range.location, 0)
        XCTAssertEqual(range.length, result.length)
    }
}


Playground/AttributedPlayground.playground/Contents.swift
//: Playground - noun: a place where people can play

import AppKit
import Foundation
import Attributed

//: All the ways you can create an Attributed:

//: No formatting at all
let noFormatting = Attributed()

//: Apply a color
let redText = Attributed(color: NSColor.redColor())
let blueText = Attributed(color: NSColor.blueColor())

//: Apply a font
let boldText = Attributed(font: NSFont.boldSystemFontOfSize(24))

//: Multiple values
let detailsText = Attributed(attributes: [NSForegroundColorAttributeName: NSColor.greenColor(), NSFontAttributeName: NSFont.systemFontOfSize(18)])

//: Usage once you have an Attributed:

//: Apply the formatting to a String
let stringWithAttributes = redText.toString { "Hi mom!" }
stringWithAttributes

//: Apply the formatting to a NSAttributedString
let attributedStringWithAttributes = redText.toString { NSAttributedString(string: "The quick brown fox") }
attributedStringWithAttributes

//: Combining Strings
let multipleStrings = boldText.combine(strings: "Hi mom!", "The quick brown fox")
multipleStrings

//: Nesting
let nestedStrings = redText.toString { boldText.toString { "Header: " } + "here are the details" }
nestedStrings

//: More nesting
let anotherNestedString = redText.toString { boldText.toString { "Header: " } + detailsText.toString { "here are the details" } + blueText.toString{ "@" } }
anotherNestedString

//: Strings and attributed strings can be added together
let addedString = stringWithAttributes + " How are you?"
addedString
